### PR TITLE
imputation_sklearn optimization

### PR DESCRIPTION
**Pull request by dattranm**

### CODE REVIEW

File: src/main/anovos/data_transformer/transformers.py [hunk 1]
```diff
@@ -1763,6 +1802,8 @@ def imputation_sklearn(
 
     """
 
+    if persist:
+        idf = idf.persist(persist_option)
```
[the remark this answers is not in the document]
Comment (VishnuGowthem) [reply src/main/anovos/data_transformer/transformers.py:1806]: this is more generically for persist_option - not at this point in code I mean.

File: src/main/anovos/data_transformer/transformers.py [hunk 2]
```diff
@@ -1676,17 +1677,24 @@ def imputation_MMM(
...
     spark,
     idf,
-    list_of_cols="missing",
-    drop_cols=[],
-    method_type="KNN",
-    sample_size=500000,
-    pre_existing_model=False,
-    model_path="NA",
-    output_mode="replace",
+    list_of_cols: list = "missing",
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/data_transformer/transformers.py]: I removed the suggested `list` type for `list_of_cols` Vishnu, and add a bit more description to explain this.
For the second point, across the repo, in `imputation_MMM` function, we also use the same `list_of_cols="missing"`. I think we should stick with this for now to standardize the argument.

File: src/main/anovos/data_transformer/transformers.py [hunk 3]
```diff
@@ -1731,9 +1740,39 @@ def imputation_sklearn(
...
+    sample_seed
+        If use_sampling is True, this argument is used to determine the seed of sampling method.
+        (Default value = 42)
+    persist
+        Boolean argument - True or False. This argument is used to determine whether to persist on
+        binning result of source and target dataset, True will enable the use of persist, otherwise False.
+        It is recommended to set this as True for large datasets. (Default value = True)
+    persist_option
+        If persist is True, this argument is used to determine the type of persist.
+        (Default value = pyspark.StorageLevel.MEMORY_AND_DISK)
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/data_transformer/transformers.py]: I have added the list of the pyspark.StorageLevel from pyspark official documentation in here.
For the label, according to Killian comment on #268 ,https://github.com/anovos/anovos/pull/268#discussion_r948473258, it would be better to just set it to pyspark.StorageLevel.MEMORY_AND_DISK to avoid confusion.

File: src/main/anovos/data_transformer/transformers.py [hunk 4]
```diff
@@ -1851,10 +1892,22 @@ def imputation_sklearn(
...
+                    fraction=sample_size / count_idf,
+                    method_type=sample_method,
+                    stratified_type=stratified_type,
+                    seed_value=sample_seed,
+                )
+                if persist:
+                    idf_model = idf_model.persist(persist_option)
+        else:
+            idf_model = idf
         idf_pd = idf_model.select(list_of_cols).toPandas()
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/data_transformer/transformers.py]: Resolved with https://github.com/anovos/anovos/pull/337/commits/402964046aaa383a06f71224499800d1a37f7f89

File: src/main/anovos/data_transformer/transformers.py [hunk 5]
```diff
@@ -1676,17 +1677,24 @@ def imputation_MMM(
...
-    output_mode="replace",
+    list_of_cols: list = "missing",
+    drop_cols: list = [],
+    method_type: str = "regression",
+    use_sampling: bool = True,
+    sample_method: str = "random",
+    strata_cols: str = "all",
+    stratified_type: str = "population",
+    sample_size: int = 10000,
+    sample_seed: int = 42,
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/data_transformer/transformers.py]: got it Vishnu. I will keep it like this for now, as you can see the same usage of `data_sample` in `correlation_matrix` function.
https://github.com/anovos/anovos/blob/a833099c9ec6dc2ef3ee68e2fca638bc4eaa8379/src/main/anovos/data_analyzer/association_evaluator.py#L40

File: src/main/anovos/data_transformer/transformers.py [hunk 6]
```diff
@@ -1731,9 +1740,39 @@ def imputation_sklearn(
         a few handful of them. (Default value = [])
     method_type
         "KNN", "regression".
-        "KNN" option trains a sklearn.impute.KNNImputer. "regression" option trains a sklearn.impute.IterativeImputer (Default value = "KNN")
+        "KNN" option trains a sklearn.impute.KNNImputer. "regression" option trains a sklearn.impute.IterativeImputer
+        (Default value = "regression")
+    use_sampling
+        Boolean argument - True or False. This argument is used to determine whether to use random sample method on
+        source and target dataset, True will enable the use of sample method, otherwise False.
```
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/data_transformer/transformers.py]: Resolved with https://github.com/anovos/anovos/commit/402964046aaa383a06f71224499800d1a37f7f89

File: src/main/anovos/data_transformer/transformers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dattranm) [reply src/main/anovos/data_transformer/transformers.py:1813]: In here, whenever `pyspark.StorageLevel` receive the wrong value, the pyspark library has its own builtin error check for this:

**Comment by dattranm** [comment]
Thank @ranjanravish and @VishnuGowthem for the comment and review! I have addressed all of your comments, with some comments of my own that can lead to further discussion. Can you guys have a quick look again and let me know if there is anything else I need to address.

File: src/main/anovos/data_transformer/transformers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (VishnuGowthem) [reply src/main/anovos/data_transformer/transformers.py]: okay noted.

**Review by VishnuGowthem — APPROVED** [review]
I am good with the changes subject to all workflows passing, thanks @dattranm !

**Comment by VishnuGowthem** [comment]
> I am good with the changes subject to all workflows passing, thanks @dattranm !

I think last set of workflows has passed. So I am good to merge this. Thanks !

**Comment by dattranm** [comment]
Thanks @VishnuGowthem for your review and approval. @kajanansangar can you have a final look and review on this to confirm everything is ready to merge?

**Review by kajanansangar — APPROVED** [review]
